### PR TITLE
vSphere: External CCM translation, Zonal 

### PR DESCRIPTION
**Pull request by jcpowermac**
This PR converts existing INI from in-tree
to external CCM INI.

Also adds the INI section for zonal labels
if FailureDomains is defined and has a length
other than 0.

### CODE REVIEW

**Comment by jcpowermac** [comment]
Still continuing on this PR.  Need to work on node networking section and then tests.

**Comment by jcpowermac** [comment]
@JoelSpeed @rvanderp3 can you take a look when you have a moment, thanks!

File: pkg/cloud/vsphere/vsphere.go [hunk 1]
```diff
@@ -89,3 +95,204 @@ func NewProviderAssets(config config.OperatorConfig) (common.CloudProviderAssets
...
+				datacenters := strings.Join(vcenter.Datacenters[:], ",")
+				_, err = setOrGetKeyValue(vCenterSection, "datacenters", datacenters, true)
+				if err != nil {
+					return "", err
+				}
+			}
+		}
+
+		if infra.Spec.PlatformSpec.VSphere.FailureDomains != nil {
+			if len(infra.Spec.PlatformSpec.VSphere.FailureDomains) != 0 {
```
Comment (JoelSpeed) [inline pkg/cloud/vsphere/vsphere.go:155]: Is the first of these if statements actually required?

File: pkg/cloud/vsphere/vsphere.go [hunk 2]
```diff
@@ -89,3 +95,204 @@ func NewProviderAssets(config config.OperatorConfig) (common.CloudProviderAssets
...
+
+				err = setLabelsSection(cfg)
+				if err != nil {
+					return "", err
+				}
```
Comment (JoelSpeed) [inline pkg/cloud/vsphere/vsphere.go:160]: Inline the error decl/check and wrap the error coming out please

File: pkg/cloud/vsphere/vsphere.go [hunk 3]
```diff
@@ -89,3 +95,204 @@ func NewProviderAssets(config config.OperatorConfig) (common.CloudProviderAssets
...
+				return err
+			}
+		}
+	}
+
+	// if there are no new or existing keys on [Nodes] section
+	// [Nodes] should not exist, delete section
+	if len(nodesSection.Keys()) == 0 {
+		cfg.DeleteSection("Nodes")
+	}
```
Comment (JoelSpeed) [inline pkg/cloud/vsphere/vsphere.go:266]: Should this go before the previous loop? And return early?

File: pkg/cloud/vsphere/vsphere.go [hunk 4]
```diff
@@ -89,3 +95,204 @@ func NewProviderAssets(config config.OperatorConfig) (common.CloudProviderAssets
...
+	return nil
+}
+
+func setOrGetSection(cfg *ini.File, name string) (*ini.Section, error) {
+	if cfg.HasSection(name) {
+		klog.Infof("%s section found...", name)
+		return cfg.GetSection(name)
+	} else {
+		klog.Infof("Creating new section: %s", name)
+		return cfg.NewSection(name)
```
Comment (JoelSpeed) [inline pkg/cloud/vsphere/vsphere.go:277]: Logs here are probably too chatty at `V(0)`, this is pretty debug level IMO

File: pkg/cloud/vsphere/vsphere.go [hunk 5]
```diff
@@ -89,3 +95,204 @@ func NewProviderAssets(config config.OperatorConfig) (common.CloudProviderAssets
...
+	}
+
+	if infra.Spec.PlatformSpec.VSphere != nil {
+		err = setNodesSection(cfg, &infra.Spec.PlatformSpec.VSphere.NodeNetworking)
+		if err != nil {
+			return "", err
+		}
+
+		if infra.Spec.PlatformSpec.VSphere.VCenters != nil {
+			if len(infra.Spec.PlatformSpec.VSphere.VCenters) != 0 {
```
[the remark this answers is not in the document]
Comment (jcpowermac) [reply pkg/cloud/vsphere/vsphere.go]: Nope, I was being cautious, removed

File: pkg/cloud/vsphere/vsphere.go [hunk 6]
```diff
@@ -89,3 +95,204 @@ func NewProviderAssets(config config.OperatorConfig) (common.CloudProviderAssets
...
+				// delete all existing [VirtualCenter] sections
+				// they will be recreated below
+				for _, section := range cfg.Sections() {
+					if strings.Contains(section.Name(), "VirtualCenter") {
+						cfg.DeleteSection(section.Name())
+					}
+				}
```
[the remark this answers is not in the document]
Comment (jcpowermac) [reply pkg/cloud/vsphere/vsphere.go]: I was thinking the infra object is the single source of truth and the 3cmo sets that
If that is wrong I will certainly change it

File: pkg/cloud/vsphere/vsphere.go (hunk 3 above)
Comment (jcpowermac) [reply pkg/cloud/vsphere/vsphere.go:268]: Maybe you have a suggestion here. 
I need to know if the node networking struct has any values assigned. I couldn't find an easy/quick of checking all struct parameters so I figured loop over the map, if there are no ini section keys I know no values were set in the struct and just delete that section.

File: pkg/cloud/vsphere/vsphere.go (hunk 4 above)
Comment (jcpowermac) [reply pkg/cloud/vsphere/vsphere.go]: fixed

**Comment by jcpowermac** [comment]
fyi @gnufied

**Comment by jcpowermac** [comment]
I will work on this tomorrow morning to get back in shape

File: pkg/cloud/vsphere/vsphere_test.go [hunk 7]
```diff
@@ -59,3 +70,338 @@ func TestResourcesRenderingSmoke(t *testing.T) {
...
+secret-namespace = "kube-system"
+insecure-flag = "1"
+
+[Workspace]
+server = "vcenter.sddc-44-236-21-251.vmwarevmc.com"
+datacenter = "SDDC-Datacenter"
+default-datastore = "WorkloadDatastore"
+folder = "/SDDC-Datacenter/vm/jcallen"
+
+[VirtualCenter "vcenter.sddc-44-236-21-251.vmwarevmc.com"]
```
Comment (lobziik) [inline pkg/cloud/vsphere/vsphere_test.go:260]: Is this virtual center expected to be removed from resulted config?

Comment (jcpowermac) [reply pkg/cloud/vsphere/vsphere_test.go]: I changed the test to use more generic names. The `[VirtualCenter]` section should remain. The only section removed will be `[Workspace]`.

Comment (jcpowermac) [reply pkg/cloud/vsphere/vsphere_test.go]: @lobziik I double checked too, purposefully made the structs unequal to make sure the test worked. 
I probably should add some test that fail.

**Pull request closed without merging**